File: backend_samples/library/javascript/AzureAIVisionFaceDeviceAttestation/src/ios/parsers.ts
/**
 * Decoders for the two CBOR payloads Apple's App Attest APIs return, plus the
 * credCert nonce-extension extractor.
 *
 *   parseAppAttestToken      — DCAppAttestService.attestKey → AppAttestObject
 *   parseAppAttestAssertion  — DCAppAttestService.generateAssertion → AppAttestAssertionObject
 *   extractNonceFromCredCert — pulls the nonce OCTET STRING out of credCert
 *                              extension OID 1.2.840.113635.100.8.2
 */
import { Constructed, fromBER, OctetString, Sequence } from 'asn1js';
import { z } from 'zod';
import { getExtensionValue } from '../cert_utils';
import { AUTH_DATA_HEADER_BYTES, FLAGS_OFFSET, NONCE_OID, RP_ID_HASH_BYTES, SIGN_COUNT_OFFSET } from './constants';
import { decodeCbor } from './cbor';
import type {
  AppAttestObject,
  AppAttestAssertionObject,
} from './types';

const byteStringSchema = z.custom<Buffer>(Buffer.isBuffer, { error: 'expected a CBOR byte string' });
const cborMapSchema = z.instanceof(Map).transform(value => Object.fromEntries(value));
const appAttestSchema = cborMapSchema.pipe(z.object({
  fmt: z.string(),
  authData: byteStringSchema,
  attStmt: cborMapSchema.pipe(z.object({
    x5c: z.tuple([byteStringSchema, byteStringSchema]).rest(z.unknown()),
    receipt: byteStringSchema.nullable().catch(null),
  })),
}));
const appAssertionSchema = cborMapSchema.pipe(z.object({
  signature: byteStringSchema,
  authenticatorData: byteStringSchema.refine(value => value.length >= AUTH_DATA_HEADER_BYTES, {
    error: `assertion authenticatorData must contain at least ${AUTH_DATA_HEADER_BYTES} bytes`,
  }),
}));

/**
 * Decode the base64-CBOR attestation token returned by
 * DCAppAttestService.attestKey. Validates the outer shape Apple documents
 * (fmt + attStmt{x5c, receipt?} + authData) and pulls out the two certs from
 * x5c (credCert + intermediate).
 */
export function parseAppAttestToken(tokenB64: string): AppAttestObject {
  const tokenBuf = Buffer.from(tokenB64, 'base64');
  const [topRaw] = decodeCbor(tokenBuf);
  const { fmt, authData, attStmt: { x5c: [credCertDer, intermediateDer], receipt } } = appAttestSchema.parse(topRaw);
  return { fmt, authData, credCertDer, intermediateDer, receiptLength: receipt?.length ?? 0, receipt };
}

/**
 * Decode the base64-CBOR assertion returned by
 * DCAppAttestService.generateAssertion. Validates signature + authenticatorData
 * are present and that authenticatorData is at least 37 bytes (rpIdHash || flags
 * || signCount).
 */
export function parseAppAttestAssertion(assertionB64: string): AppAttestAssertionObject {
  const buf = Buffer.from(assertionB64, 'base64');
  const [topRaw] = decodeCbor(buf);
  return appAssertionSchema.parse(topRaw);
}

/**
 * Decode the first 37 bytes of an assertion's `authenticatorData`:
 *   rpIdHash (32) || flags (1) || signCount (4)
 *
 * Used by `verifyAssertionAgainstAuthCert` (replaces hand-rolled offset reads)
 * and by `inspectAppAttest` for display.
 */
export function parseAssertionAuthData(authenticatorData: Buffer): {
  rpIdHash: Buffer;
  flags: number;
  signCount: number;
} {
  if (authenticatorData.length < AUTH_DATA_HEADER_BYTES) {
    throw new Error(`authenticatorData is ${authenticatorData.length} bytes, < ${AUTH_DATA_HEADER_BYTES}`);
  }
  return {
    rpIdHash: authenticatorData.subarray(0, RP_ID_HASH_BYTES),
    flags: authenticatorData[FLAGS_OFFSET],
    signCount: authenticatorData.readUInt32BE(SIGN_COUNT_OFFSET),
  };
}

/**
 * Walk into the DER bytes of the credCert and pull out the OCTET STRING value
 * embedded in the App Attest nonce extension (OID 1.2.840.113635.100.8.2).
 * Structure:
 *   Extension ::= SEQUENCE { extnID OID, critical BOOLEAN DEFAULT FALSE, extnValue OCTET STRING }
 *   extnValue (DER) = SEQUENCE { [1] EXPLICIT OCTET STRING nonce }
 */
export function extractNonceFromCredCert(credCertDer: Buffer): Buffer | null {
  const container = getExtensionValue(credCertDer, NONCE_OID);
  if (!container) return null;
  try {
    const decoded = fromBER(container);
    if (decoded.offset !== container.length || !(decoded.result instanceof Sequence)) return null;
    const fields = decoded.result.valueBlock.value;
    const tagged = fields[0];
    if (fields.length !== 1 || !(tagged instanceof Constructed) ||
        tagged.idBlock.tagClass !== 3 || tagged.idBlock.tagNumber !== 1) return null;
    const [nonce] = tagged.valueBlock.value;
    if (tagged.valueBlock.value.length !== 1 || !(nonce instanceof OctetString) || nonce.idBlock.isConstructed) return null;
    return Buffer.from(nonce.valueBlock.valueHexView);
  } catch {
    return null;
  }
}

File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/ios/parsers.py
"""
Decoders for the two CBOR payloads Apple's App Attest APIs return, plus the
credCert nonce-extension extractor.
"""

from __future__ import annotations

from typing import Optional

import cbor2
from asn1crypto import core

from ..base64_utils import decode_base64
from ..cert_utils import get_extension_value
from .constants import (
    AUTH_DATA_HEADER_BYTES,
    FLAGS_OFFSET,
    NONCE_OID,
    RP_ID_HASH_BYTES,
    SIGN_COUNT_OFFSET,
)
from .types import AppAttestAssertionObject, AppAttestObject


class _NonceExtension(core.Sequence):
    _fields = [("nonce", core.OctetString, {"explicit": 1})]


def parse_app_attest_token(token_b64: str) -> AppAttestObject:
    buf = decode_base64(token_b64)
    top = cbor2.loads(buf)
    if not isinstance(top, dict):
        raise ValueError("top-level CBOR is not a map")

    fmt = top.get("fmt")
    att_stmt = top.get("attStmt")
    auth_data = top.get("authData")
    if not isinstance(fmt, str):
        raise ValueError("missing fmt")
    if not isinstance(att_stmt, dict):
        raise ValueError("missing attStmt")
    if not isinstance(auth_data, (bytes, bytearray)):
        raise ValueError("missing authData")

    x5c = att_stmt.get("x5c")
    if not isinstance(x5c, list) or len(x5c) < 2:
        raise ValueError("attStmt.x5c must have >= 2 certs")
    cred_cert_der = x5c[0]
    intermediate_der = x5c[1]
    if not isinstance(cred_cert_der, (bytes, bytearray)) or not isinstance(intermediate_der, (bytes, bytearray)):
        raise ValueError("attStmt.x5c entries must be byte strings")

    receipt_raw = att_stmt.get("receipt")
    receipt_length = len(receipt_raw) if isinstance(receipt_raw, (bytes, bytearray)) else 0

    return AppAttestObject(
        fmt=fmt,
        authData=bytes(auth_data),
        credCertDer=bytes(cred_cert_der),
        intermediateDer=bytes(intermediate_der),
        receiptLength=receipt_length,
    )


def parse_app_attest_assertion(assertion_b64: str) -> AppAttestAssertionObject:
    buf = decode_base64(assertion_b64)
    top = cbor2.loads(buf)
    if not isinstance(top, dict):
        raise ValueError("assertion top-level CBOR is not a map")

    signature = top.get("signature")
    authenticator_data = top.get("authenticatorData")
    if not isinstance(signature, (bytes, bytearray)):
        raise ValueError("assertion missing signature")
    if not isinstance(authenticator_data, (bytes, bytearray)):
        raise ValueError("assertion missing authenticatorData")
    if len(authenticator_data) < AUTH_DATA_HEADER_BYTES:
        raise ValueError(f"assertion authenticatorData is {len(authenticator_data)} bytes, < {AUTH_DATA_HEADER_BYTES}")

    return AppAttestAssertionObject(
        signature=bytes(signature),
        authenticatorData=bytes(authenticator_data),
    )


def parse_assertion_auth_data(authenticator_data: bytes) -> dict:
    if len(authenticator_data) < AUTH_DATA_HEADER_BYTES:
        raise ValueError(f"authenticatorData is {len(authenticator_data)} bytes, < {AUTH_DATA_HEADER_BYTES}")
    rp_id_hash = authenticator_data[:RP_ID_HASH_BYTES]
    flags = authenticator_data[FLAGS_OFFSET]
    sign_count = int.from_bytes(authenticator_data[SIGN_COUNT_OFFSET:AUTH_DATA_HEADER_BYTES], "big")
    return {"rpIdHash": rp_id_hash, "flags": flags, "signCount": sign_count}


def extract_nonce_from_cred_cert(cred_cert_der: bytes) -> Optional[bytes]:
    """Pull the OCTET STRING value inside the App Attest
    nonce extension (OID 1.2.840.113635.100.8.2).
    """
    container = get_extension_value(cred_cert_der, NONCE_OID)
    if container is None:
        return None
    try:
        fields = _NonceExtension.load(container, strict=True)
        if len(fields) != 1:
            return None
        nonce = fields["nonce"].native
        return nonce if fields.dump(force=True) == container else None
    except (ValueError, TypeError):
        return None
